web/packages/teleport/src/Account/Account.tsx

```diff
@@ -180,13 +186,203 @@ export function Account({
+    0x00001809: 'Irish',
+    0x0000201a: 'Bosnian (Cyrillic)',
+    0x00010401: 'Arabic (102)',
+    0x00010402: 'Bulgarian (Latin)',
+    0x00010405: 'Czech (QWERTY)',
+    0x00010407: 'German (IBM)',
+    0x00010408: 'Greek (220)',
+    0x00010409: 'United States - Dvorak',
+    0x0001040a: 'Spanish Variation',
+    0x0001040e: 'Hungarian 101-key',
+    0x00010410: 'Italian (142)',
+    0x00010415: 'Polish (214)',
+    0x00010416: 'Portuguese (Brazilian ABNT2)',
+    0x00010418: 'Romanian (Standard)',
+    0x00010419: 'Russian (Typewriter)',
+    0x0001041b: 'Slovak (QWERTY)',
+    0x0001041e: 'Thai Pattachote',
+    0x0001041f: 'Turkish F',
+    0x00010426: 'Latvian (QWERTY)',
+    0x00010427: 'Lithuanian',
+    0x0001042b: 'Armenian Western',
+    0x0001042e: 'Sorbian Extended',
+    0x0001042f: 'Macedonian (FYROM) - Standard',
+    0x00010437: 'Georgian (QWERTY)',
+    0x00010439: 'Hindi Traditional',
+    0x0001043a: 'Maltese 48-key',
+    0x0001043b: 'Sami Extended Norway',
+    0x00010445: 'Bengali - INSCRIPT (Legacy)',
+    0x0001045a: 'Syriac Phonetic',
+    0x0001045b: 'Sinhala - wij 9',
+    0x0001045d: 'Inuktitut - Naqittaut',
+    0x00010465: 'Divehi Typewriter',
+    0x0001080c: 'Belgian (Comma)',
+    0x0001083b: 'Finnish with Sami',
+    0x00011009: 'Canadian Multilingual Standard',
+    0x00011809: 'Gaelic',
+    0x00020401: 'Arabic (102) AZERTY',
+    0x00020402: 'Bulgarian (phonetic layout)',
+    0x00020405: 'Czech Programmers',
+    0x00020408: 'Greek (319)',
+    0x00020409: 'United States - International',
+    0x00020418: 'Romanian (Programmers)',
+    0x0002041e: 'Thai Kedmanee (non-ShiftLock)',
+    0x00020422: 'Ukrainian (Enhanced)',
+    0x00020427: 'Lithuanian New',
+    0x00020437: 'Georgian (Ergonomic)',
+    0x00020445: 'Bengali - INSCRIPT',
+    0x0002083b: 'Sami Extended Finland-Sweden',
+    0x00030402: 'Bulgarian (phonetic layout)',
+    0x00030408: 'Greek (220) Latin',
+    0x00030409: 'United States-Devorak for left hand',
+    0x0003041e: 'Thai Pattachote (non-ShiftLock)',
+    0x00040408: 'Greek (319) Latin',
+    0x00040409: 'United States-Dvorak for right hand',
+    0x00050409: 'Greek Latin',
+    0x00060408: 'Greek Polytonic',
+  };
+
+  let layout = ctx.preferences.keyboardLayout;
+  let value = { label: layouts[layout], value: layout };
+  let options = Object.keys(layouts).map(k => {
+    return { label: layouts[k], value: parseInt(k) };
+  });
+
   return (
     <Relative>
       <FeatureBox>
         <FeatureHeader>
           <FeatureHeaderTitle>Account Settings</FeatureHeaderTitle>
         </FeatureHeader>
         <Flex flexDirection="column" gap={4}>
+          <Box>
+            <SingleRowBox>
+              <Header
+                title="Keyboard layout"
+                description="Keyboard layout used by Windows Desktop sessions"
+                icon={<Icon.Edit />}
+                actions={
+                  <Box width="210px">
+                    <Select
```

Comment:
We should handle loading and error states, I'd do it in this way:

```
Subject: [PATCH] Add loading and error states
---
Index: web/packages/teleport/src/Account/Account.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/web/packages/teleport/src/Account/Account.tsx b/web/packages/teleport/src/Account/Account.tsx
--- a/web/packages/teleport/src/Account/Account.tsx	(revision 8c8a32e7cdb1045b18110545ef3007dfd5753dcc)
+++ b/web/packages/teleport/src/Account/Account.tsx	(date 1715158437138)
@@ -26,8 +26,10 @@
 import { useStore } from 'shared/libs/stores';
 
 import { SingleRowBox } from 'design/MultiRowBox';
+import Alert from 'design/Alert';
 
 import Select from 'shared/components/Select';
+import { useAsync } from 'shared/hooks/useAsync';
 
 import useTeleport from 'teleport/useTeleport';
 import {
@@ -349,6 +351,10 @@
     return { label: layouts[k], value: parseInt(k) };
   });
 
+  const [updateKeyboardLayoutAttempt, runUpdateKeyboardLayout] = useAsync(
+    ctx.updatePreferences
+  );
+
   return (
     <Relative>
       <FeatureBox>
@@ -358,6 +364,12 @@
         <Flex flexDirection="column" gap={4}>
           <Box>
             <SingleRowBox>
+              {updateKeyboardLayoutAttempt.status === 'error' && (
+                <Alert kind="danger">
+                  Failed to update the keyboard layout:{' '}
+                  {updateKeyboardLayoutAttempt.statusText}
+                </Alert>
+              )}
               <Header
                 title="Keyboard layout"
                 description="Keyboard layout used by Windows Desktop sessions"
@@ -369,12 +381,13 @@
                         if (Array.isArray(selected)) {
                           selected = selected[0];
                         }
-                        ctx
-                          .updatePreferences({
-                            keyboardLayout: selected.value,
-                          })
-                          .then(() => (value = selected));
+                        runUpdateKeyboardLayout({
+                          keyboardLayout: selected.value,
+                        });
                       }}
+                      isDisabled={
+                        updateKeyboardLayoutAttempt.status === 'processing'
+                      }
                       value={value}
                       options={options}
                     />

```

Reply:
That makes sense, thank you very much!